### ISSUE TITLE
rust.vim loading broken under dein 

### ISSUE THREAD

**Issue by mqudsi** [Shougo/dein.vim#221]
**Warning:  I will close the issue without the minimal init.vim and the reproduce ways.**

# Problems summary
Very similar to #110, dein problems loading rust.vim persist until now. Since the time when I commented on #110, dein would sometimes randomly refuse to load the rust.vim plugin correctly but the problem would disappear when I would update neovim, dein, and rust.vim

This is the first time I'm able to reproduce it each item against latest.

## Expected
rust.vim should load, syntax should be highlighted, commands such as `:Rustfmt` should be available.

## Environment Information (Required!)

 * dein.vim version(SHA1):
9c892cbcbf016dbaea4bbdb9a0c2acd45329264d

 * OS: WSL

 * Vim/neovim version:

```
NVIM v0.2.1-442-g69f0847
Build type: Release
Compilation: /usr/bin/cc -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/home/mqudsi/random/neovim/build/config -I/home/mqudsi/random/neovim/src -I/home/mqudsi/random/neovim/.deps/usr/include -I/home/mqudsi/random/neovim/.deps/usr/include -I/home/mqudsi/random/neovim/.deps/usr/include -I/home/mqudsi/random/neovim/.deps/usr/include -I/home/mqudsi/random/neovim/.deps/usr/include -I/home/mqudsi/random/neovim/.deps/usr/include -I/usr/include -I/home/mqudsi/random/neovim/build/src/nvim/auto -I/home/mqudsi/random/neovim/build/include
Compiled by mqudsi@ZBook
```

## Provide a minimal .vimrc with less than 50 lines (Required!)

```vim
"dein Scripts-----------------------------
if &compatible
        set nocompatible               " Be iMproved
endif

set runtimepath+=$HOME/.config/nvim/dein/repos/github.com/Shougo/dein.vim
filetype off

if dein#load_state('$HOME/.config/nvim/dein/')
        call dein#begin('$HOME/.config/nvim/dein/')
        call dein#add('Shougo/dein.vim')

        call dein#add('rust-lang/rust.vim', { 'on_ft': ['rs'], 'rev': 'b77ac8ecbd4baaa23cca612e1c9b3df5ea23da9b' })  

        call dein#end()
        call dein#save_state()
endif

filetype plugin indent on
syntax enable

autocmd BufRead,BufNewFile FileType rust compiler cargo

" If you want to install not installed plugins on startup.
if dein#check_install()
  call dein#install()
endif
```

## The reproduce ways from Vim starting (Required!)

 1. nvim src/main.rs 
 2. try calling `:Rustfmt`

**Comment by mqudsi** [comment]
OK, this is driving me nuts. Even though my dein was already at the latest version, starting nvim and then running `call dein#update()` made everything work again. This is what usually happens.

What does `dein#update()` do that could cause a plugin to go from not working to working when dein and the plugin are both already up-to-date (i.e. no new releases or code changes)?

**Comment by mqudsi** [comment]
I reverted my original `init.vim` instead of the minimal one I posted above and the problem re-appeared. No syntax highlighting for rust files, no rust provided by rust.vim. I run `call dein#update()`, exited vim, started it again, no change. Highlighting still doesn't work.

**Comment by mqudsi** [comment]
OK, here's the deal: I had `on_ft` set to `rs` which is incorrect; that's the default rust filetype but not the filetype installed by rust.vim, which is `rust`.

Once `on_ft` is set incorrectly, removing `on_ft` does not restore functioning behavior until `dein#update()` is called. 

Does that sound plausible?

**Comment by Shougo** [comment]
Yes, rs filetype is invalid.  So this is feature.

> Does that sound plausible?

I will fix it.

**Comment by mqudsi** [comment]
I just ran into another bug, I'm not sure if it's the same as this one or not.

I added the line `call dein#add('tomtom/tcomment_vim')` to my init.vim and started neovim. `dein` detected the change and installed `tcomment`

However, no tcomment features worked at all and there was no sign of it being installed aside from the help documents being listed under `:help tc<TAB>`, for example, `:TComment` did not exist. I added and removed the plugin a few times but that made no difference. Then I ran `call dein#update()` and TComment started working.
